### PR TITLE
0722 dynamic topics

### PR DESCRIPTION
**Pull request by zmstone**
Support dynamic topics for supervised producers.
Call `wolff:ensure_supervised_dynamic_producers(ClientId, #{group => GroupName, ...})` to start a group-producer with no topics added in advance.
And call `wolff:send2` or `wolff:send_sync2` to publish messages with topic provided as an argument.

### CODE REVIEW

File: src/wolff_producers_sup.erl [hunk 1]
```diff
@@ -61,3 +62,16 @@ child_spec(ClientId, ProducerId, Config) ->
...
    }.
+
+%% Find the running process (gen_server of wolff_producers) from thie child ID.
+get_producers_pid(ID) ->
+  Children = supervisor:which_children(?SUPERVISOR),
+  case lists:keyfind(ID, 1, Children) of
+    {ID, Pid, _Type, _Modules} when is_pid(Pid) ->
+      Pid;
+    Other ->
+      throw(#{cause => producers_not_found_uder_supervisor,
```
Comment (thalesmg) [inline src/wolff_producers_sup.erl:73]:
```suggestion
      throw(#{cause => producers_not_found_under_supervisor,
```

File: src/wolff_producers.erl [hunk 2]
```diff
@@ -34,7 +36,15 @@
 -include("wolff.hrl").
 -include_lib("stdlib/include/ms_transform.hrl").
 
--opaque producers() :: topic_producers() | linked_topic_producers().
+-opaque producers() :: dynamic_topic_producers() | topic_producers() | linked_topic_producers().
+
+%% When calling APIs with dynmic producers, topic must be preented as function argument.
```
Comment (thalesmg) [inline src/wolff_producers.erl:41]:
```suggestion
%% When calling APIs with dynamic producers, topic must be presented as function argument.
```

File: src/wolff_producers.erl [hunk 3]
```diff
@@ -20,6 +20,8 @@
 -export([start_linked_producers/3, stop_linked/1]).
 -export([start_supervised/3, stop_supervised/1, stop_supervised/2]).
 -export([pick_producer/2, lookup_producer/2, cleanup_workers_table/1]).
+%% Dynamic topics
+-export([start_supervised_dynamic/2, pick_producer2/3]).
```
Comment (thalesmg) [inline src/wolff_producers.erl:24]:
```suggestion
-export([start_supervised_dynamic/2, pick_producer2/3]).
-export([add_topic/2, remove_topic/2]).
```
?

File: src/wolff_producers.erl [hunk 4]
```diff
@@ -186,16 +246,45 @@ pick_producer(#{workers := Workers,
...
+pick_producer2(#{client_id := ClientId,
+                 group := Group,
+                 topic := T0
+                } = Producers, Topic, Batch) ->
+  T0 =/= ?DYNAMIC andalso error("cannot_add_topic_to_non_dynamic_producer"),
+  Count0 = get_partition_cnt(ClientId, Group, Topic),
+  Count = resove_partitions_cnt(ClientId, Group, Topic, Count0),
+  pick_producer3(Producers, Topic, Batch, Count).
+
+resove_partitions_cnt(_ClientId, _Group, _Topic, C) when is_integer(C) andalso C > 0 ->
```
Comment (thalesmg) [inline src/wolff_producers.erl:265]:
```suggestion
resolve_partitions_cnt(_ClientId, _Group, _Topic, C) when is_integer(C) andalso C > 0 ->
```

File: src/wolff_producers.erl [hunk 5]
```diff
@@ -164,18 +204,38 @@ stop_supervised(ClientId, ?NS_TOPIC(NS, Topic) = ID) ->
...
+-spec throw_unknown(_, _, _, _, _) -> no_return().
+throw_unknown(ClientId, Group, Topic, Reason, RespFrom) ->
+  throw(#{client_id => ClientId,
+          group => Group,
+          topic => Topic,
+          cause => Reason,
+          response => RespFrom
+         }).
+
+add_topic(Group, Topic) ->
```
Comment (thalesmg) [inline src/wolff_producers.erl:229]: Should we also have an API to remove topics from group?

File: src/wolff_producers.erl [hunk 6]
```diff
@@ -186,16 +246,45 @@ pick_producer(#{workers := Workers,
...
+resove_partitions_cnt(ClientId, Group, Topic, ?partition_count_unavailable) ->
+  resove_partitions_cnt(ClientId, Group, Topic, ?UNKNOWN(0));
+resove_partitions_cnt(ClientId, Group, Topic, ?UNKNOWN(Since)) ->
+  case (now_ts() - Since) > timer:seconds(?unknown_topic_cache_expire_seconds) of
+    true ->
+      case add_topic(Group, Topic) of
+        ok ->
+          get_partition_cnt(ClientId, Group, Topic);
+        {error, unknown_topic_or_partition} ->
+          throw_unknown(ClientId, Group, Topic, unknown_topic_or_partition, received);
```
Comment (thalesmg) [inline src/wolff_producers.erl:276]:
```suggestion
          throw_unknown(ClientId, Group, Topic, unknown_topic_or_partition, refreshed);
```
?

File: src/wolff_producers.erl [hunk 7]
```diff
@@ -343,11 +446,24 @@ handle_cast(Cast, St) ->
...
+    case init_one_topic_producers(St, Topic, _Attempts = 0) of
+      ok ->
+        {#{Topic => ?initialized}, ok};
+      discard ->
+        ok = mark_topic_unknown(Group, Topic),
+        {#{}, {error, unknown_topic_or_partition}};
+      {error, ?not_initialized(_, Reason) = TopicStatus1} ->
+        {TopicStatus1, {error, Reason}}
+    end,
+  {St#{producers_status => maps:merge(Status, TopicStatus)}, Reply}.
```
Comment (thalesmg) [inline src/wolff_producers.erl:464]: Nit:

```suggestion
  {St#{producers_status := maps:merge(Status, TopicStatus)}, Reply}.
```

File: test/wolff_dynamic_topics_tests.erl [hunk 8]
```diff
@@ -0,0 +1,135 @@
+-module(wolff_dynamic_topics_tests).
+
+-include("wolff.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("kafka_protocol/include/kpro.hrl").
+
+-define(KEY, key(?FUNCTION_NAME)).
+-define(HOSTS, [{"localhost", 9092}]).
+
+%% Checks that having different producers to the same kafka topic works.
```
Comment (thalesmg) [inline test/wolff_dynamic_topics_tests.erl:10]: Nit: the test doesn't seem to actually test 2 different dynamic producer groups targeting the same topic(s).

File: test/wolff_dynamic_topics_tests.erl [hunk 9]
```diff
@@ -103,6 +102,35 @@ bad_producers_test() ->
   ?assertError("cannot_add_topic_to_non_dynamic_producer", wolff:send_sync2(Producers, <<"test-topic">>, [Msg], 1000)),
   ok.
 
+topic_add_remove_test() ->
```
Comment (thalesmg) [inline test/wolff_dynamic_topics_tests.erl:105]: Nit: `wolff:remove` is not tested here.

**Review by thalesmg — APPROVED** [review]

File: test/wolff_dynamic_topics_tests.erl (hunk 9 above)
Comment (zmstone) [reply test/wolff_dynamic_topics_tests.erl:106]: right. fixed. thanks

**Review by thalesmg — APPROVED** [review]

**Pull request merged**